dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -44,45 +44,164 @@ namespace DB
+    using Base::Base;
+    MPPDataPacketPtr readForLocal();
+};
+
+using TunnelSenderPtr = std::shared_ptr<TunnelSender>;
+using SyncTunnelSenderPtr = std::shared_ptr<SyncTunnelSender>;
+using AsyncTunnelSenderPtr = std::shared_ptr<AsyncTunnelSender>;
+using LocalTunnelSenderPtr = std::shared_ptr<LocalTunnelSender>;
+
 /**
- * MPPTunnelBase represents the sender of an exchange connection.
- *
- * (Deprecated) It is designed to be a template class so that we can mock a MPPTunnel without involving gRPC.
+ * MPPTunnel represents the sender of an exchange connection.
  *
- * The lifecycle of a MPPTunnel can be indicated by `connected` and `finished`:
- * | Stage                          | `connected` | `finished` |
- * |--------------------------------|-------------|------------|
- * | After constructed              | false       | false      |
- * | After `close` before `connect` | false       | true       |
- * | After `connect`                | true        | false      |
- * | After `consumerFinish`         | true        | true       |
+ * The lifecycle of a MPPTunnel can be indicated by TunnelStatus:
+ * | Previous Status        | Event           | New Status             |
+ * |------------------------|-----------------|------------------------|
+ * | NaN                    | Construction    | Unconnected            |
+ * | Unconnected            | Close           | Finished               |
+ * | Unconnected            | Connection      | Connected              |
+ * | Connected              | WriteDone       | WaitingForSenderFinish |
+ * | Connected              | Close           | WaitingForSenderFinish |
+ * | Connected              | Encounter error | WaitingForSenderFinish |
+ * | WaitingForSenderFinish | Sender Finished | Finished               |
  *
- * To be short: before `connect`, only `close` can finish a MPPTunnel; after `connect`, only `consumerFinish` can.
+ * To be short: before connect, only close can finish a MPPTunnel; after connect, only Sender Finish can.
  *
- * Each MPPTunnel has a consumer to consume data. There're two kinds of consumers: local and remote.
- * - Remote consumer is owned by MPPTunnel itself. MPPTunnel will create a thread and run `sendLoop`.
- * - Local consumer is owned by the associated ExchangeReceiver (in the same process).
+ * Each MPPTunnel has a Sender to consume data. There're three kinds of senders: sync_remote, local and async_remote.
  * 
- * The protocol between MPPTunnel and consumer:
+ * The protocol between MPPTunnel and Sender:
  * - All data will be pushed into the `send_queue`, including errors.
- * - MPPTunnel may close `send_queue` to notify consumer normally finish.
- * - Consumer may close `send_queue` to notify MPPTunnel that an error occurs.
- * - After `connect` only the consumer can set `finished` to `true`.
- * - Consumer's state is saved in `consumer_state` and be available after consumer finished.
+ * - MPPTunnel may close `send_queue` to notify Sender normally finish.
```

Comment:
There are two ways to `close` send_queue: `finish` and `cancel`, is there any difference between the two `close`?

Reply:
Change 'close' to 'finish'.